src/deepmol/datasets/datasets.py

```diff
@@ -574,28 +591,23 @@ def merge(self, datasets: List[Dataset]) -> 'NumpyDataset':
         y = self.y
         ids = self.ids
         mols = self.mols
-        flag2 = False
 
         for ds in datasets:
-            mols = np.append(mols, ds.mols, axis=0)
-            y = np.append(y, ds.y, axis=0)
-            ids = np.append(ids, ds.ids, axis=0)
-            if X is not None:
-                if len(X[0]) == len(ds.X[0]):
-                    X = np.append(X, ds.X, axis=0)
-                else:
-                    flag2 = False
+            ids = merge_arrays(ids, len(mols), ds.ids, len(ds.mols))
+            if len(set(ids)) != len(ids):
+                raise ValueError(f'IDs must be unique! IDs are {ids}')
+            y = merge_arrays(y, len(mols), ds.y, len(ds.mols))
+            if X is None or ds.X is None:
+                print('Features are not the same length/type... Recalculate features for all inputs!')
+                X = None
+            elif len(X.shape) == 1 and len(ds.X.shape) == 1:
+                X = merge_arrays(X, len(mols), ds.X, len(ds.mols))
             else:
-                flag2 = False
-        if flag2:
-            print('Features are not the same length/type... '
-                  '\nRecalculate features for all inputs! '
-                  '\nAppending empty array in dataset features!')
-            return NumpyDataset(mols, None, y, ids)
-        else:
-            return NumpyDataset(mols, X, y, ids, self.features2keep)
+                X = merge_arrays_of_arrays(X, ds.X)
+            mols = np.append(mols, ds.mols, axis=0)
+        return NumpyDataset(mols, X, y, ids, self.features2keep)
 
-    def save_to_csv(self, path):
+    def save_to_csv(self, path: str):
```

Comment:
Personal taste only, but I prefer **to_csv** as in **pandas**

Reply:
Yeah, I agree.